"""
Start the optimized gRPC server on port 50052 for testing
"""

import asyncio
import grpc
from concurrent import futures
import time
import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import optimized_lipsyncsrv_pb2
import optimized_lipsyncsrv_pb2_grpc
from optimized_grpc_server import OptimizedLipSyncServicer
from optimized_inference_engine import optimized_engine

async def serve():
    """Start the gRPC server on port 50052"""
    
    print("=" * 80)
    print("üöÄ STARTING OPTIMIZED gRPC SERVER ON PORT 50052")
    print("=" * 80)
    print("   ‚ö° gRPC with HTTP/2")
    print("   ‚ö° Protocol Buffers serialization")
    print("   ‚ö° Pre-loaded videos in RAM")
    print("   ‚ö° Memory-mapped audio features")
    print("   ‚ö° Zero I/O overhead")
    print("   ‚ö° GPU Batch Processing")
    print("   ‚ö° Audio Batch Optimization")
    print("=" * 80)
    
    # Auto-load optimized sanders model
    print("\nüì¶ Loading optimized sanders...")
    try:
        result = await optimized_engine.load_package(
            "sanders",
            "models/sanders"  # Path relative to minimal_server directory
        )
        
        if result["status"] == "success":
            print("\n‚úÖ sanders loaded successfully!")
            print(f"   Frame count: {result['frame_count']}")
            print(f"   Initialization time: {result['initialization_time_s']:.2f}s")
            print(f"   Device: {result['device']}")
            print(f"   Videos loaded: {', '.join(result['videos_loaded'])}")
            print(f"   Audio features shape: {result['audio_features_shape']}")
            print(f"   Memory-mapped audio: {result['memory_mapped_audio']}")
        else:
            print(f"‚ö†Ô∏è Model load failed: {result.get('error', 'Unknown error')}")
            return
            
    except Exception as e:
        print(f"‚ùå Error loading sanders: {e}")
        import traceback
        traceback.print_exc()
        return
    
    # Create gRPC server
    server = grpc.aio.server(
        futures.ThreadPoolExecutor(max_workers=10),
        options=[
            ('grpc.max_send_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.so_reuseport', 1),
            ('grpc.use_local_subchannel_pool', 1),
        ]
    )
    
    # Add servicer
    optimized_lipsyncsrv_pb2_grpc.add_OptimizedLipSyncServiceServicer_to_server(
        OptimizedLipSyncServicer(),
        server
    )
    
    # Bind to port 50052
    listen_addr = '[::]:50052'
    server.add_insecure_port(listen_addr)
    
    print("\n" + "=" * 80)
    print(f"‚úÖ Server listening on {listen_addr}")
    print("=" * 80)
    print("\nüéØ Available RPC Methods:")
    print("   ‚Ä¢ GenerateFrame - Single frame generation")
    print("   ‚Ä¢ GenerateBatchInference - GPU batch processing")
    print("   ‚Ä¢ GenerateBatchWithAudio - Audio batch optimization")
    print("   ‚Ä¢ LoadModel - Dynamic model loading")
    print("   ‚Ä¢ ListModels - List loaded models")
    print("   ‚Ä¢ HealthCheck - Server health status")
    print("=" * 80)
    print("\nüìä Press Ctrl+C to stop the server\n")
    
    # Start server
    await server.start()
    
    # Wait for termination
    try:
        await server.wait_for_termination()
    except KeyboardInterrupt:
        print("\n\nüõë Shutting down server...")
        await server.stop(grace=5)
        print("‚úÖ Server stopped gracefully")

if __name__ == '__main__':
    asyncio.run(serve())
